tests/signals/tests.py

```diff
@@ -341,6 +341,38 @@ class Created(models.Model):
         Created()
         self.assertEqual(received, [])
 
+    @isolate_apps('signals', kwarg_name='apps')
+    def test_disconnect_return_value(self, apps):
+        """
+        Signal.disconnect() return value should be always True or False,
+        independently whether the sender uses the Model or its string
+        reference.
+        """
+
+        class Created(models.Model):
+            pass
+
+        def receiver(**kwargs):
+            pass
+
+        sender = Created
+        signals.post_init.connect(receiver, sender=sender, apps=apps)
+        self.assertTrue(
+            signals.post_init.disconnect(receiver, sender=sender, apps=apps)
+        )
+        self.assertFalse(
+            signals.post_init.disconnect(receiver, sender=sender, apps=apps)
+        )
+
+        sender = 'signals.Created'
+        signals.post_init.connect(receiver, sender=sender, apps=apps)
+        self.assertTrue(
+            signals.post_init.disconnect(receiver, sender=sender, apps=apps)
+        )
+        self.assertFalse(
+            signals.post_init.disconnect(receiver, sender=sender, apps=apps)
+        )
+
```

Comment:
I think we can add two assertions to the `test_disconnect` instead of a new test case, e.g.
```diff
diff --git a/tests/signals/tests.py b/tests/signals/tests.py
index e0a24ccc6e..ed7c124932 100644
--- a/tests/signals/tests.py
+++ b/tests/signals/tests.py
@@ -333,7 +333,8 @@ class LazyModelRefTests(BaseSignalSetup, SimpleTestCase):
             received.append(kwargs)
 
         signals.post_init.connect(receiver, sender='signals.Created', apps=apps)
-        signals.post_init.disconnect(receiver, sender='signals.Created', apps=apps)
+        self.assertIs(signals.post_init.disconnect(receiver, sender='signals.Created', apps=apps), True)
+        self.assertIs(signals.post_init.disconnect(receiver, sender='signals.Created', apps=apps), False)
 
         class Created(models.Model):
             pass
```

Reply:
I started with an isolated test to illustrate the issue clearly.

I can make the change if you think it doesn't affect the readability of `test_disconnect`.
One drawback I can think of is that this way we are only testing the case where we use the string and not the class. 

I am happy to move forward with either solution.

Reply:
`disconnect()` works without this patch when sender is a model class. Extra tests are welcome, but we should add them in a separate commit/PR, e.g.
```diff
diff --git a/tests/signals/tests.py b/tests/signals/tests.py
index e0a24ccc6e..11094ebaad 100644
--- a/tests/signals/tests.py
+++ b/tests/signals/tests.py
@@ -326,7 +326,27 @@ class LazyModelRefTests(BaseSignalSetup, SimpleTestCase):
             signals.post_init.disconnect(self.receiver, sender=Created)
 
     @isolate_apps('signals', kwarg_name='apps')
-    def test_disconnect(self, apps):
+    def test_disconnect_model(self, apps):
+        class Created(models.Model):
+            pass
+
+        signals.post_init.connect(self.receiver, sender=Created, apps=apps)
+        try:
+            self.assertIs(
+                signals.post_init.disconnect(self.receiver, sender=Created, apps=apps),
+                True,
+            )
+            self.assertIs(
+                signals.post_init.disconnect(self.receiver, sender=Created, apps=apps),
+                False,
+            )
+            Created()
+            self.assertEqual(self.received, [])
+        finally:
+            signals.post_init.disconnect(self.receiver, sender=Created, apps=apps)
+
+    @isolate_apps('signals', kwarg_name='apps')
+    def test_disconnect_lazy_ref(self, apps):
         received = []
 
         def receiver(**kwargs):

```

Reply:
To split the tests in two is a good idea: one to test the call with the model and another to test the call with the lazy ref. I shall follow your suggestion and make a new PR with the new tests after we get this sorted out.

Meanwhile, I am struggling with the lazy ref test. In your example above, the `Created` model is only declared after the `connect() / disconnect()` steps. After digging into this, I believe the model is not registered when we call `disconnect()`.

I pushed the modifications so you can have a look. 
Let me know if this is getting off-topic. We can focus in the problem at hands (`disconnect()` return value) and leave the model order for another issue.

Reply:
We should push tests that works without this patch in advance, so I moved it to a separate PR, see #14205.